discovery/spec.md

```diff
@@ -778,54 +778,51 @@ upon return of the `202 Accepted` from the original request.
-Services are to be added or updated in the Discovery Endpoint.
-
-The Services in the request MUST be processed in the order in which they are
-given. While any error during the update MUST cause the rollback of any effect
-of the request, the processing of each Service in the request MUST be
-applied as if all previous Services in the request were already completed.
-This means that attribute uniqueness checking, such as on the `name`
-attribute, MUST be done on the new state of all Services as if all previously
-specified Services in the incoming request were already successfully processed.
-
-For example, if a Discovery Endpoint has a Service called `MyService`, it is
-possible to rename that Service to `YourService` and then to add, or update,
-another Service to use `MyService` as its `name`.
-
-Likewise, while it might be less than optimal, it is possible for
-a request to update a Service multiple times if the same `id` is used more
-than once within a request.
+contain all Services in the request or fail. The body of the request message
+MUST contain an array of zero or more Service entries.
+
+The following rules apply to processing the Services specified in the request:
+- any error processing the request MUST result in the entire operation failing
+  and no updates to the Discovery Endpoint are to happen.
+- a Service (as identified by its `id`) MUST NOT appear more than once in the
+  request.
+- any existing Service (as identified by its `id`) MUST be completely
+  replaced by the Service definition in the request, except as noted below.
+  This means that any missing property in the incoming request MUST be
+  treated as a request to delete the property from the existing Service
+  definition.
+- if the incoming Service has an `epoch` value then the request MUST fail if
+  that value is not greater than the current Service's `epoch` value.
+- if the incoming Service does not have an `epoch` value then the Service's
+  `epoch` value MUST be set to a valid value of the Discovery Endpoint's
+  choosing. This means that if there is an existing Service with the same
+  `id` then the value MUST be set to a value greater than the current value.
+- if the incoming Service does not contain an `id` then the Discovery Endpoint
+  MUST assign an appropriate value. The value MUST be globally unique.
+- aside from `id`, `epoch` and `url`, unless there is an out of band agreement,
+  all mandatory attributes MUST be present in the request, and a Discovery
+  Endpoint MUST reject requests that are missing such attributes.
+- the Discovery Endpoint MUST ensure that the `url` of the resulting Service
+  is correct even if that means overriding the value from the request.
+
+There is no requirement that the incoming Services be processed in the
+order in which they appear. However, any constraint/consistency checks
+(e.g. `name` uniqueness checking) MUST only be performed after all of the
+incoming Services have been processed.
```

Comment:
true - is there a better wording?

Reply:
always a good question... maybe:
```suggestion
There is no requirement that the incoming Services be processed in the
order in which they appear. Constraints apply only to the result of 
the request.  For example, if a Service with `name: "dog"` exists
and a request creates a new Service with `name: "dog"` and later in the
request updates the existing Service to have `name: "cat"` the request
is expected to succeed even if the Discovery Endpoint implementation
would cause two services to temporarily have `name: "dog"`.
```
The example is long so I would be fine if it were just omitted.

Reply:
I tweaked it slightly - see watcha think

Reply:
The new wording is great.

Reply:
The new wording is great.